Tests/AutoCollection/Heartbeat.tests.ts

```diff
@@ -0,0 +1,249 @@
+            heartbeat3.enable(true, client.config);
+            HeartBeat.INSTANCE.enable(true, client.config);
+            const stub3 = sinon.stub(heartbeat3["_client"], "trackMetric");
+            const requestURL = "http://169.254.169.254"
+            nock(requestURL, {
+                reqheaders: {
+                    "Metadata": "true"
+                }
+            })
+            .get('/metadata/instance/compute?api-version=2017-12-01&format=json')
+            .reply(
+                200,
+                {
+                    'vmId': 1,
+                    'subscriptionId': 2,
+                    'osType': 'Linux'
+                }
+            );
+
+            heartbeat3["trackHeartBeat"](client.config, () => {
+                assert.equal(stub3.callCount, 1, "calls trackMetric for the VM heartbeat metric");
+                assert.equal(stub3.args[0][0].name, "HeartBeat", "uses correct name for heartbeat metric");
+                assert.equal(stub3.args[0][0].value, 0, "checks value is 0");
+                const keys3 = Object.keys(stub3.args[0][0].properties);
+                assert.equal(keys3.length, 4, "makes sure 4 kv pairs are added when resource type is VM");
+                assert.equal(keys3[0], "sdk", "sdk is added as a key");
+                assert.equal(keys3[1], "azInst_vmId",  "azInst_vmId is added as a key");
+                assert.equal(keys3[2], "azInst_subscriptionId", "azInst_subscriptionId is added as a key");
+                assert.equal(keys3[3], "azInst_osType", "azInst_osType is added as a key");
+
+                const properties3 = stub3.args[0][0].properties;
+                assert.equal(properties3["sdk"], Context.sdkVersion, "sdk version is read from Context");
+                assert.equal(properties3["azInst_vmId"], 1, "azInst_vmId is read from response");
+                assert.equal(properties3["azInst_subscriptionId"], 2, "azInst_subscriptionId is read from response");
+                assert.equal(properties3["azInst_osType"], "Linux", "azInst_osType is read from response");
+
+                stub3.restore();
+                heartbeat3.dispose();
+                done();
+            });
+        });
+
+        it("should read correct VM information from response", (done) => {
+            const heartbeat4: HeartBeat = new HeartBeat(client);
+            heartbeat4.enable(true, client.config);
+            HeartBeat.INSTANCE.enable(true, client.config);
+            const stub4 = sinon.stub(heartbeat4["_client"], "trackMetric");
+
+            heartbeat4["_vmData"] = {
+                vmId: "1",
+                subscriptionId: "2",
+                osType: "Linux"
+            };
+            heartbeat4["_isVM"] = true;
+            // sinon.stub(Util, "makeRequest").returns(new fakeResponse(false));
+            sinon.stub(Util, 'makeRequest', (config: any, requestUrl: string, requestOptions: any, requestCallback: Function) => {
```

Comment:
Without this stub, the trackHeartBeat will call _getAzureComputeMetadata which calls Util.makeRequest that will assign _isVM = false (reset what I set on line 172) so the assertions won't pass. This stub here is trying to force makeRequest return a fakeresponse with statusCode=200. But the way I code here will report error as "statusCode undefined". @markwolff @MSNev Can you please take a look. I think heartbeat4 test is more straightforward than heartbeat3 test... Thanks a lot! I'm stuck on this test for too long... 😩

Reply:
Using the useFakeRequest() to avoid needing to mock like this -- plenty of examples in 1P/3P Js code bases.